JSON loading had been added

Y-Pazniak opened the pull request:
Also i've added parents filling from json
I've added json loading.

Y-Pazniak committed:
I've added json loading.
Also i've added parents filling from json

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/tree/HelperTree.java:
```diff
@@ -31,39 +36,58 @@ public static Map<StepsIndicator, String> fillFieldsToCommandsMap() {
-        otherPhysical.addChildren(List.of(otherPhysicalAge));
+        BufferedReader bufferedReader = null;
+        try {
+            bufferedReader = new BufferedReader(new FileReader("/home/eugene/IdeaProjects/custom_calculator_chatbot/src/main/resources/tree.json"));
```
Please, use try-with-resources

MaximPilipeiko commented on pom.xml:
```diff
@@ -63,6 +63,12 @@
+
+        <dependency>
+            <groupId>com.google.code.gson</groupId>
+            <artifactId>gson</artifactId>
```
Not a call to action, just a notion. Jackson is a little bit more popular than gson and is used in popular frameworks like Spring

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/service/BundleResourcesServant.java:
```diff
@@ -28,11 +28,11 @@ public String getString(final String stringToGetFromBundle) {
-            File fileResources = new File("D:\\IdeaProjects\\custom_calculator_bot\\src\\main\\resources");
-            URL[] resourcesFileUrls = {fileResources.toURI().toURL()};
-            ClassLoader loader = new URLClassLoader(resourcesFileUrls);
+            final File fileResources = new File("D:\\IdeaProjects\\custom_calculator_bot\\src\\main\\resources");
```
Using an absolute path is a bad practice. Please, use relevant path like.
this topic could help you https://stackoverflow.com/questions/17351043/how-to-get-absolute-path-to-file-in-resources-folder-of-your-project

MaximPilipeiko commented on src/main/resources/tree.json:
```diff
@@ -0,0 +1,85 @@
+{
+  "parent": null,
```
you can rid "parent": null. If field doesn't exist, it's null.

Y-Pazniak committed:
GSON was replaced by Jackson.
Missing try-with-resources replaced by File
Absolute path was replaced.
parent: null in json was removed

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/tree/HelperTree.java:
```diff
@@ -31,39 +36,35 @@ public static Map<StepsIndicator, String> fillFieldsToCommandsMap() {
-        Node otherPhysical = new Node(other, Command.PHYSICAL_PERSON);
-        Node otherPhysicalAge = new Node(otherPhysical, Command.AGE);
-        otherPhysical.addChildren(List.of(otherPhysicalAge));
+        final ObjectMapper objectMapper = new ObjectMapper();
```
It could be a static class field. You haven't to create it on every method call.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/tree/HelperTree.java:
```diff
@@ -31,39 +36,35 @@ public static Map<StepsIndicator, String> fillFieldsToCommandsMap() {
+        try {
+            treeRootJson = objectMapper.readValue(file, Node.class);
+        } catch (final IOException e) {
+            e.printStackTrace();
```
I think we have to stop the application if cannot read the tree. Please, throw exception on higher level.

MaximPilipeiko commented on src/main/resources/tree.json:
```diff
@@ -0,0 +1,70 @@
+          "key": "/physical",
+          "children": [
+            {
+              "key": "age",
```
Why this key value without `/`

MaximPilipeiko reviewed (changes requested):
Lets disscuss deserialisation on a call

Y-Pazniak committed:
JSON descriptions were added
If tree reading fails, bot shuts down.
Command replaced by Enum, but still doesn't work with json.

Y-Pazniak committed:
Json and Enum are friends now. Commands from json are synchronized with enum Command.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/controller/Bot.java:
```diff
@@ -35,7 +37,8 @@ public String getBotUsername() {
     public String getBotToken() {
         return
-                System.getenv("custom_by_utilsbor_bot");
+                "6490861114:AAHVFOpQANIn_EQr1A1PcLrCMNzVlGkJIig";
```
Why do you hardcode value?

Y-Pazniak replied:
Another solutions don't work on my Ubuntu system by unknown reason.
I use ubuntu documentation to set EnvVariable:
Terminal: export BOT=TESTBOT
Checking: echo $BOT 
Everything is fine. 

Then I follow Java documentation: use System.getenv("MY_VARIABLE"); to take variable.
And it returns null. 

I don't know why it happens and how to fix it 

![image](https://github.com/Y-Pazniak/custom_calculator_chatbot/assets/73634869/23285907-f33d-4401-acfb-4b3b6d7917d5)

![image](https://github.com/Y-Pazniak/custom_calculator_chatbot/assets/73634869/67c24efd-44ce-46e9-b31b-d622639090f1)

MaximPilipeiko replied:
What about System.getProperty.
https://stackoverflow.com/questions/7054972/java-system-properties-and-environment-variables

Y-Pazniak replied:
It's amazing! Everything works fine, thanks a lot!
If you want to set your own Properties in Idea: three points (icon in the top right corner) - edit - modify options - add VM options - set pair key-value (example: -Dtest=test).

Also I have just read about the difference between properties and env variables and now I understand better why System.properties are better (they work only for JVM and do not affect on OS)

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/controller/MessageRouter.java:
```diff
@@ -35,38 +37,60 @@ public SendMessage route(final Update update) {
-            case Command.START -> {
+        final String answer;
+
+        while (true) {
```
Why infinite loop is needed?

Y-Pazniak replied:
We had switch-case here before. 
Now it becomes impossible to use switch-case, so I replaced it by if.
If i use if-else, it is a terrible spagetti-code. So I decided to use while and brakes to avoid too long if-else to get to the SorryMessage: 
![image](https://github.com/Y-Pazniak/custom_calculator_chatbot/assets/73634869/91b3d4da-9041-47f4-bcbf-7f24589fddf9)

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/controller/MessageRouter.java:
```diff
@@ -35,38 +37,60 @@ public SendMessage route(final Update update) {
+        final String answer;
+
+        while (true) {
+            if (Objects.equals(usersMessage, Command.START.getCommand())) {
```
From my view, an infinite loop looks more stylish.

Y-Pazniak replied:
sorry, don't understand you here - need explanation

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/tree/HelperTree.java:
```diff
@@ -2,68 +2,73 @@
+
     public static Map<StepsIndicator, String> fillFieldsToCommandsMap() {
-        Map<StepsIndicator, String> fieldsToCommands = new HashMap<>(14);
+        final Map<StepsIndicator, String> fieldsToCommands = new HashMap<>(14);
```
Is it possible to use Command as a value instead of String?
```suggestion
        final Map<StepsIndicator, Command> fieldsToCommands = new HashMap<>(14);
```

Y-Pazniak replied:
Commands are better than Strings, you explained me why. 
Replaced all Strings to Commands

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/tree/HelperTree.java:
```diff
@@ -2,68 +2,73 @@
+        try {
+            file = Paths.get(res.toURI()).toFile();
+        } catch (URISyntaxException e) {
+            e.printStackTrace();
```
Please throw `TreeReadingException`. It's impossible to handle this exception gracefully.

Y-Pazniak replied:
I tried, but it demands URISyntaxException - it doesn't work even if I inherit my own exception from URISyntaxException
![image](https://github.com/Y-Pazniak/custom_calculator_chatbot/assets/73634869/7b5d99d8-5708-4390-a823-58e3525d3cc5)

MaximPilipeiko replied:
Throw, not catch ;)
YOu have to catch URISyntaxException and throw TreeReadingException

Y-Pazniak replied:
Fixed, thank you

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/tree/Node.java:
```diff
@@ -1,14 +1,19 @@
-    private final Node parent;
+    private Node parent;
     private final List<Node> children;
     private final String key;
```
It's better to use enum instead of String.
You can look [deserialization node example](https://github.com/MaximPilipeiko/jackson/blob/master/src/main/java/org/example/Main.java)

Y-Pazniak replied:
Fixed it. I don't know what happened - but now it works.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/exception/TreeReadingException.java:
```diff
@@ -0,0 +1,12 @@
+
+import java.io.IOException;
+
+public class TreeReadingException extends IOException {
```
This name gives more information: what tree, is it our service exception or pure java exception?
```suggestion
public class UrilsborCommandTreeReadingException extends IOException {
```

Y-Pazniak replied:
Renamed, thanks

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/service/UserProgressManager.java:
```diff
@@ -33,93 +35,123 @@ public void createNewUserProgress(final String chatID) throws UtilsborException
-
+        final String message;
+
+        if (Objects.equals(requestingCommand, Command.EAES.getCommand())) {
```
It's betteruse type `Command` for the argument `requestingCommand`  instead of casting `Command` to `String`.

Y-Pazniak replied:
Fixed it according to your advice, thank you

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/service/UserProgressManager.java:
```diff
@@ -33,93 +35,123 @@ public void createNewUserProgress(final String chatID) throws UtilsborException
+
+        if (Objects.equals(requestingCommand, Command.EAES.getCommand())) {
+            userProgress.setCountryOrigin(CountryOrigin.EAES);
+        } else {
```
```suggestion
        } else if (Objects.equals(requestingCommand, Command.OTHER_COUNTRIES.getCommand())) {
            userProgress.setCountryOrigin(CountryOrigin.OTHER);
        }
```

Y-Pazniak replied:
We have an agreement always use {} after if and else, even if it is not necessary.
I removed {} but probably we need to discuss it during our call

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/service/UserProgressManager.java:
```diff
@@ -33,93 +35,123 @@ public void createNewUserProgress(final String chatID) throws UtilsborException
-            case Command.VOLUME_MORE_3500_CM -> userProgress.setVolumeOfEngine(VolumeOfEngine.MORE_3500);
+        final String message;
+
+        while (true) {
```
switch case is better. And why loop is needed is not clear.

Y-Pazniak replied:
It was impossible to use switch-case because I used Strings - it is forbidden to use String in switch-case. 
Using string here was a mistake, I replaced it by switch-case.

Infinity loop was needed here because it made code more beautiful - I hadn't use long 
if(){
}else{
if() {
}else {
if()}}

MaximPilipeiko commented on src/main/resources/words_ru.properties:
```diff
@@ -37,3 +37,5 @@ answers.file_reading_error = Извините, произошла ошибка 
 
 answers.invalid_order_exception = Извините, недостаточно данных для расчета или нарушена очередность. Попробуйте начать с команды /start или свяжитесь с разработчиками
+
+answers.tree_reading_error = impossible to read the tree 
```
just 'tree' is too abstract
```suggestion
answers.tree_reading_error = impossible to read the command tree 
```

Y-Pazniak replied:
Fixed it, thank you

Y-Pazniak committed:
Now this code works with Command from Node.

Y-Pazniak committed:
Useless code has been removed

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/controller/MessageRouter.java:
```diff
@@ -35,36 +35,47 @@ public SendMessage route(final Update update) {
-                    Command.VOLUME_BETWEEN_3000_3500_CM, Command.VOLUME_MORE_3500_CM ->
-                    answer = userProgressManager.processEngineVolume(usersMessage, chatID);
+
+            case EAES, OTHER_COUNTRIES -> answer = userProgressManager.processCarOrigin(requestingCommand, chatID);
```
What about `break`? Is it not needed?

Y-Pazniak replied:
Thanks for asking! I have googled it and discovered this.

1. If we are using traditional switch-case block like 
switch (smth) {
case 1: 
           dosmth(1;) 
           break;
case 2: 
           dosmth(2); 
           break;
case 3:
   dosmth(3);
   break;
}
Using break; is NECESSARY to avoid doing useless code.

2. If we use arrows like 
switch(smth) {
case 1 -> dosmth(1);
case 2 -> dosmth(2);
case 1 -> dosmth(3);
}
Using break; is NOT necessary. 

I didn't know about it.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/constants/Command.java:
```diff
@@ -1,33 +1,71 @@
+    }
+
+    public static Command getCommandByKey(final String key){
+        for (Command command1 : Command.values()) {
```
`command1` is not enterprise like naming

Y-Pazniak replied:
Fixed

Y-Pazniak committed:
Hardcoded token had been removed.
Tree Reading Error had been added correctly.
break in switch-case with our syntax (with arrows ->) is useless
Proper name for Command command variable had been added

Y-Pazniak committed:
Added two empty branches.
M1 branch works with new branches - need to fill another branches.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/controller/Bot.java:
```diff
@@ -24,6 +24,7 @@ public static Bot getBot() {
     private Bot() {
         //initDB();
         messageRouter = MessageRouter.getInstance();
+        CommandTree.getInstance();
```
Please, add a comment on why you call getInstance without initializing any variable like in the line above.

Y-Pazniak replied:
Comment was added

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/controller/MessageRouter.java:
```diff
@@ -35,36 +35,49 @@ public SendMessage route(final Update update) {
+
+            case PHYSICAL_PERSON, JURIDICAL_PERSON ->
+                    answer = userProgressManager.processOwnerType(requestingCommand, chatID);
+
```
too many empty lines to my taste. It's not call to action

Y-Pazniak replied:
Agree, the code had been cleaned
